File: source/pong_env.py
# pong_env.py
import numpy as np
import random
import sys

class PongEnv:
    """
    Pong environment with 5D state representation.
    Includes improved collision detection logic.
    """
    def __init__(self, width=600, height=400, paddle_height=60, paddle_width=10, ball_radius=8):
        # --- Game Dimensions ---
        self.width = width
        self.height = height
        self.paddle_height = paddle_height
        self.paddle_width = paddle_width
        self.ball_radius = ball_radius

        # --- Normalization Helpers ---
        self.norm_width = self.width / 2.0
        self.norm_height = self.height / 2.0

        # --- Game Objects State ---
        self.ball_x = 0.0; self.ball_y = 0.0
        self.ball_vx = 0.0; self.ball_vy = 0.0
        self.paddle1_y = 0.0 # Agent paddle (left)
        self.paddle2_y = 0.0 # Opponent paddle (right)

        # --- Physics/Game Parameters ---
        self.paddle_speed = 12.0
        self.ball_speed_initial = 10.0
        self.ball_speed_increase = 0.4
        self.current_ball_speed = self.ball_speed_initial
        self.max_ball_speed = 24.0
        self.opponent_speed_limit = 12.0

        # --- RL Interface ---
        self.state_dim = 5
        self.action_dim = 3

        # --- Pygame Rendering Attributes ---
        self.screen = None
        self.clock = None

        self.reset()

    def _normalize_state(self):
        """ Normalizes the current game state for the agent (5 dimensions). """
        state = [
            self.ball_x / self.norm_width,
            self.ball_y / self.norm_height,
            self.ball_vx / self.max_ball_speed,
            self.ball_vy / self.max_ball_speed,
            self.paddle1_y / self.norm_height,
        ]
        return np.array(state, dtype=np.float32)

    def reset(self):
        """ Resets the environment to the start of a new game. """
        self.ball_x = 0.0
        self.ball_y = random.uniform(-self.norm_height * 0.5, self.norm_height * 0.5)
        angle = random.uniform(-np.pi / 4, np.pi / 4)
        direction = 1 if random.random() < 0.5 else -1
        self.current_ball_speed = self.ball_speed_initial
        self.ball_vx = direction * self.current_ball_speed * np.cos(angle)
        self.ball_vy = self.current_ball_speed * np.sin(angle)
        self.paddle1_y = 0.0
        self.paddle2_y = 0.0
        observation = self._normalize_state()
        info = {}
        return observation, info

    def step(self, action):
        """ Executes one time step within the environment. """
        # 1. Move Agent Paddle
        if action == 1: self.paddle1_y -= self.paddle_speed
        elif action == 2: self.paddle1_y += self.paddle_speed
        paddle_half_height = self.paddle_height / 2.0
        self.paddle1_y = np.clip(self.paddle1_y, -self.norm_height + paddle_half_height, self.norm_height - paddle_half_height)

        # 2. Move Opponent Paddle
        target_y = self.paddle2_y
        if self.ball_y > self.paddle2_y:
             move_amount = min(self.opponent_speed_limit, self.ball_y - self.paddle2_y)
             target_y += move_amount
        elif self.ball_y < self.paddle2_y:
             move_amount = min(self.opponent_speed_limit, self.paddle2_y - self.ball_y)
             target_y -= move_amount
        self.paddle2_y = np.clip(target_y, -self.norm_height + paddle_half_height, self.norm_height - paddle_half_height)

        # --- Store previous ball position for collision check ---
        prev_ball_x = self.ball_x
        prev_ball_y = self.ball_y

        # 3. Move Ball
        self.ball_x += self.ball_vx
        self.ball_y += self.ball_vy

        # 4. Ball Collisions
        reward = 0.0
        terminated = False
        paddle_hit = False # Flag to prevent scoring if paddle hit occurs

        # --- Improved Collision Logic ---

        #   a. Top/Bottom Walls (Check first)
        if self.ball_y - self.ball_radius < -self.norm_height or self.ball_y + self.ball_radius > self.norm_height:
            self.ball_vy *= -1
            # Correct position slightly to prevent sticking
            self.ball_y = np.clip(self.ball_y, -self.norm_height + self.ball_radius, self.norm_height - self.ball_radius)

        #   b. Paddles (Check BEFORE scoring, use previous position)
        paddle1_front_edge = -self.norm_width + self.paddle_width # Right edge of paddle 1
        paddle2_front_edge = self.norm_width - self.paddle_width  # Left edge of paddle 2

        # Collision with Agent Paddle (Paddle 1)
        # Check if ball crossed the front edge plane during this step
        if self.ball_vx < 0 and \
           prev_ball_x - self.ball_radius >= paddle1_front_edge and \
           self.ball_x - self.ball_radius < paddle1_front_edge and \
           abs(self.ball_y - self.paddle1_y) < paddle_half_height + self.ball_radius: # Vertical alignment check

            self.ball_vx *= -1
            relative_intersect_y = (self.paddle1_y - self.ball_y) / paddle_half_height
            bounce_angle = relative_intersect_y * (np.pi / 3)
            self.current_ball_speed = min(self.current_ball_speed + self.ball_speed_increase, self.max_ball_speed)
            self.ball_vx = self.current_ball_speed * np.cos(bounce_angle)
            self.ball_vy = self.current_ball_speed * -np.sin(bounce_angle)
            # Place ball exactly at paddle surface to prevent tunneling further
            self.ball_x = paddle1_front_edge + self.ball_radius
            paddle_hit = True

        # Collision with Opponent Paddle (Paddle 2)
        # Check if ball crossed the front edge plane during this step
        elif self.ball_vx > 0 and \
             prev_ball_x + self.ball_radius <= paddle2_front_edge and \
             self.ball_x + self.ball_radius > paddle2_front_edge and \
             abs(self.ball_y - self.paddle2_y) < paddle_half_height + self.ball_radius: # Vertical alignment check

            self.ball_vx *= -1
            relative_intersect_y = (self.paddle2_y - self.ball_y) / paddle_half_height
            bounce_angle = relative_intersect_y * (np.pi / 3)
            self.current_ball_speed = min(self.current_ball_speed + self.ball_speed_increase, self.max_ball_speed)
            self.ball_vx = -self.current_ball_speed * np.cos(bounce_angle)
            self.ball_vy = self.current_ball_speed * -np.sin(bounce_angle)
             # Place ball exactly at paddle surface
            self.ball_x = paddle2_front_edge - self.ball_radius
            paddle_hit = True

        #   c. Scoring (Check AFTER potential paddle collisions)
        # Only score if a paddle wasn't hit in this step AND ball is out of bounds
        if not paddle_hit:
            if self.ball_x - self.ball_radius < -self.norm_width: # Opponent scores
                reward = -1.0
                terminated = True
            elif self.ball_x + self.ball_radius > self.norm_width: # Agent scores
                reward = 1.0
                terminated = True

        # --- END Improved Collision Logic ---

        # 5. Prepare return values
        observation = self._normalize_state()
        truncated = False
        info = {}
        return observation, reward, terminated, truncated, info

    def render(self, mode='human'):
        """ Renders the environment using Pygame. """
        try:
            import pygame
        except ImportError:
            print("Pygame not installed, skipping rendering. Run 'pip install pygame'")
            return

        screen_width = int(self.width)
        screen_height = int(self.height)

        # Initialize Pygame screen and clock if not already done
        if self.screen is None:
             try:
                 pygame.display.init()
                 # Check if display is available (important for servers)
                 if not pygame.display.get_init():
                     # Fallback or error if no display
                     # For simplicity, we'll just disable rendering if no display
                     print("Warning: No display available for Pygame rendering.")
                     self.screen = None
                     return
                 self.screen = pygame.display.set_mode((screen_width, screen_height))
                 pygame.display.set_caption("Pong RL")
                 self.clock = pygame.time.Clock()
             except pygame.error as e:
                 print(f"Pygame display init error: {e}. Rendering disabled.")
                 self.screen = None # Ensure we don't try again if init fails
                 return # Stop rendering if it fails

        # If screen initialization failed previously, don't try to render
        if self.screen is None:
            return

        # Handle Pygame events (like closing the window)
        try:
            for event in pygame.event.get():
                if event.type == pygame.QUIT:
                    self.close() # Clean up pygame
                    # Signal to the calling script that rendering should stop
                    raise SystemExit("Pygame window closed by user.")
        except SystemExit: # Catch the specific exit exception
             raise # Re-raise it so the calling script can handle it
        except pygame.error as e:
            print(f"Pygame event error: {e}. Stopping rendering.")
            self.close()
            return # Stop if event handling fails

        # --- Drawing Logic ---
        def to_screen_pos(x, y):
            sx = int(x + self.norm_width)
            sy = int(y + self.norm_height)
            return sx, sy

        self.screen.fill((0, 0, 0))
        paddle_half_height = self.paddle_height / 2.0
        paddle1_rect = pygame.Rect(0, 0, self.paddle_width, self.paddle_height); paddle1_rect.center = to_screen_pos(-self.norm_width + self.paddle_width / 2, self.paddle1_y); pygame.draw.rect(self.screen, (255, 255, 255), paddle1_rect)
        paddle2_rect = pygame.Rect(0, 0, self.paddle_width, self.paddle_height); paddle2_rect.center = to_screen_pos(self.norm_width - self.paddle_width / 2, self.paddle2_y); pygame.draw.rect(self.screen, (255, 255, 255), paddle2_rect)
        ball_sx, ball_sy = to_screen_pos(self.ball_x, self.ball_y); pygame.draw.circle(self.screen, (255, 255, 255), (ball_sx, ball_sy), self.ball_radius)
        center_x = screen_width // 2
        for y in range(0, screen_height, 20): pygame.draw.rect(self.screen, (100, 100, 100), (center_x - 2, y, 4, 10))
        pygame.display.flip()
        if self.clock: self.clock.tick(120) # Keep render fast if used

    def close(self):
        """ Cleans up the Pygame window. """
        if self.screen is not None:
            try:
                import pygame
                pygame.display.quit()
                pygame.quit()
            except ImportError: pass
            except Exception as e: print(f"Error closing pygame: {e}")
            finally: self.screen = None; self.clock = None

File: docs/pong_env.js
// pong_env.js
/**
 * Simple Pong Environment for JavaScript.
 * Uses 5D state (excludes opponent paddle y).
 * Includes improved collision detection.
 * Accepts a speed multiplier in the step function, applied to final movement.
 */
class PongEnvJs {
    constructor(canvas) {
        this.canvas = canvas;
        this.ctx = canvas.getContext('2d');

        // --- Game Dimensions ---
        this.width = 600; this.height = 400;
        this.paddle_height = 60; this.paddle_width = 10;
        this.ball_radius = 8;

        // --- Normalization Helpers ---
        this.norm_width = this.width / 2.0; this.norm_height = this.height / 2.0;

        // --- Game Objects State ---
        this.ball_x = 0.0; this.ball_y = 0.0;
        this.ball_vx = 0.0; this.ball_vy = 0.0;
        this.paddle1_y = 0.0; // AI agent paddle (left)
        this.paddle2_y = 0.0; // Opponent/Player paddle (right)

        // --- Physics/Game Parameters ---
        this.base_paddle_speed = 12.0;
        this.base_ball_speed_initial = 10.0;
        this.base_ball_speed_increase = 0.4;
        this.base_max_ball_speed = 24.0;
        this.current_ball_speed_factor = 1.0;

        // --- RL Interface ---
        this.state_dim = 5;
        this.action_dim = 3;

        this._internal_reset();
    }

    _internal_reset() {
        this.ball_x = 0.0;
        this.ball_y = (Math.random() - 0.5) * this.height * 0.5;
        const angle = (Math.random() - 0.5) * (Math.PI / 2);
        const direction = (Math.random() < 0.5) ? 1 : -1;
        this.current_ball_speed_factor = 1.0;
        const initial_speed = this.base_ball_speed_initial;
        this.ball_vx = direction * initial_speed * Math.cos(angle);
        this.ball_vy = initial_speed * Math.sin(angle);
        this.paddle1_y = 0.0;
        this.paddle2_y = 0.0;
    }

    _normalize_state() {
        const state = [
            this.ball_x / this.norm_width,
            this.ball_y / this.norm_height,
            this.ball_vx / this.base_max_ball_speed,
            this.ball_vy / this.base_max_ball_speed,
            this.paddle1_y / this.norm_height,
        ];
            if (state.length !== this.state_dim) {
            console.error(`State dimension mismatch! Expected ${this.state_dim}, got ${state.length}`);
        }
        return new Float32Array(state);
    }


    reset() {
        this._internal_reset();
        const observation = this._normalize_state();
        const info = {};
        return [observation, info];
    }

    // Accepts gameSpeedMultiplier
    step(ai_action, opponent_or_player_paddle_y, deltaTime = 1/60, gameSpeedMultiplier = 1.0) {

        // Apply Multiplier to Final Movement Calculation
        const dt_factor = 60 * deltaTime * gameSpeedMultiplier; // Combined factor

        // Calculate base movement amounts for this frame
        const paddle_move_base = this.base_paddle_speed * dt_factor;

        // Adjust ball velocity magnitude based on factor (not multiplier yet)
        const current_base_speed = this.base_ball_speed_initial * this.current_ball_speed_factor;
        const speed_magnitude = Math.sqrt(this.ball_vx**2 + this.ball_vy**2);
        if (speed_magnitude > 1e-6) {
                const scale = current_base_speed / speed_magnitude;
                this.ball_vx *= scale;
                this.ball_vy *= scale;
        } else if (current_base_speed > 1e-6) {
                const angle = (Math.random() - 0.5) * (Math.PI / 2);
                const direction = (Math.random() < 0.5) ? 1 : -1;
                this.ball_vx = direction * current_base_speed * Math.cos(angle);
                this.ball_vy = current_base_speed * Math.sin(angle);
        }
        // Calculate final ball movement using the adjusted velocity and dt_factor
        const dt_ball_vx = this.ball_vx * dt_factor;
        const dt_ball_vy = this.ball_vy * dt_factor;

        const prev_ball_x = this.ball_x;

        // 1. Move AI Agent Paddle
        if (ai_action === 1) { this.paddle1_y -= paddle_move_base; }
        else if (ai_action === 2) { this.paddle1_y += paddle_move_base; }
        const paddle_half_height = this.paddle_height / 2.0;
        this.paddle1_y = Math.max(-this.norm_height + paddle_half_height, Math.min(this.norm_height - paddle_half_height, this.paddle1_y));

        // 2. Update Opponent/Player Paddle
        this.paddle2_y = opponent_or_player_paddle_y;

        // 3. Move Ball
        this.ball_x += dt_ball_vx;
        this.ball_y += dt_ball_vy;

        // 4. Ball Collisions
        let reward = 0.0; let terminated = false; let paddle_hit = false;

        //   a. Top/Bottom Walls
        if (this.ball_y - this.ball_radius < -this.norm_height || this.ball_y + this.ball_radius > this.norm_height) {
            this.ball_vy *= -1;
            this.ball_y = Math.max(-this.norm_height + this.ball_radius, Math.min(this.norm_height - this.ball_radius, this.ball_y));
        }

        //   b. Paddles
        const paddle1_front_edge = -this.norm_width + this.paddle_width;
        const paddle2_front_edge = this.norm_width - this.paddle_width;

        if (this.ball_vx < 0 && prev_ball_x - this.ball_radius >= paddle1_front_edge && this.ball_x - this.ball_radius < paddle1_front_edge && Math.abs(this.ball_y - this.paddle1_y) < paddle_half_height + this.ball_radius) {
            this.ball_vx *= -1;
            const relative_intersect_y = (this.paddle1_y - this.ball_y) / paddle_half_height;
            const bounce_angle = relative_intersect_y * (Math.PI / 3);
            this.current_ball_speed_factor = Math.min(this.current_ball_speed_factor + (this.base_ball_speed_increase / this.base_ball_speed_initial), this.base_max_ball_speed / this.base_ball_speed_initial);
            const new_speed = this.base_ball_speed_initial * this.current_ball_speed_factor;
            this.ball_vx = new_speed * Math.cos(bounce_angle);
            this.ball_vy = new_speed * -Math.sin(bounce_angle);
            this.ball_x = paddle1_front_edge + this.ball_radius;
            paddle_hit = true;
        }
        else if (this.ball_vx > 0 && prev_ball_x + this.ball_radius <= paddle2_front_edge && this.ball_x + this.ball_radius > paddle2_front_edge && Math.abs(this.ball_y - this.paddle2_y) < paddle_half_height + this.ball_radius) {
            this.ball_vx *= -1;
            const relative_intersect_y = (this.paddle2_y - this.ball_y) / paddle_half_height;
            const bounce_angle = relative_intersect_y * (Math.PI / 3);
            this.current_ball_speed_factor = Math.min(this.current_ball_speed_factor + (this.base_ball_speed_increase / this.base_ball_speed_initial), this.base_max_ball_speed / this.base_ball_speed_initial);
            const new_speed = this.base_ball_speed_initial * this.current_ball_speed_factor;
            this.ball_vx = -new_speed * Math.cos(bounce_angle);
            this.ball_vy = new_speed * -Math.sin(bounce_angle);
            this.ball_x = paddle2_front_edge - this.ball_radius;
            paddle_hit = true;
        }

        //   c. Scoring
        if (!paddle_hit) {
            if (this.ball_x - this.ball_radius < -this.norm_width) { reward = -1.0; terminated = true; }
            else if (this.ball_x + this.ball_radius > this.norm_width) { reward = 1.0; terminated = true; }
        }

        // 5. Prepare return values
        const observation = this._normalize_state();
        const truncated = false;
        const info = {};
        return [observation, reward, terminated, truncated, info];
    }

    to_screen_pos(x, y) {
        const sx = x + this.norm_width;
        const sy = y + this.norm_height;
        return [sx, sy];
        }
    render() {
        const ctx = this.ctx;
        ctx.fillStyle = '#000'; ctx.fillRect(0, 0, this.canvas.width, this.canvas.height);
        ctx.fillStyle = '#FFF';
        const paddle_half_height = this.paddle_height / 2.0;
        const [p1_sx, p1_sy] = this.to_screen_pos(-this.norm_width + this.paddle_width / 2, this.paddle1_y);
        ctx.fillRect(p1_sx - this.paddle_width / 2, p1_sy - paddle_half_height, this.paddle_width, this.paddle_height);
        const [p2_sx, p2_sy] = this.to_screen_pos(this.norm_width - this.paddle_width / 2, this.paddle2_y);
        ctx.fillRect(p2_sx - this.paddle_width / 2, p2_sy - paddle_half_height, this.paddle_width, this.paddle_height);
        const [ball_sx, ball_sy] = this.to_screen_pos(this.ball_x, this.ball_y);
        ctx.beginPath(); ctx.arc(ball_sx, ball_sy, this.ball_radius, 0, Math.PI * 2); ctx.fill();
        ctx.strokeStyle = '#555'; ctx.lineWidth = 2; ctx.setLineDash([10, 10]);
        ctx.beginPath(); ctx.moveTo(this.width / 2, 0); ctx.lineTo(this.width / 2, this.height); ctx.stroke();
        ctx.setLineDash([]);
        }
}